### PR TITLE
Clean up last `ModelCheckpoint` `makedirs` call to IOPlugin

### PR DESCRIPTION
**Pull request by jjenniferdai**
## What does this PR do?

see [comment](https://github.com/PyTorchLightning/pytorch-lightning/pull/9433#discussion_r712715429):
makedirs_before

now that we have [XLACheckpointIO](https://github.com/PyTorchLightning/pytorch-lightning/pull/9972):
- clean up the last `makedirs` call in `ModelCheckpoint` to IOPlugin
- [separate PR?] PR #9433 directly removed trainer property `should_rank_save_checkpoint`. It was added back in PR #9901, though it's not used anywhere, I think just from a merging issue. This time should we remove or deprecate it?
-- if deprecate, definitely separate PR
- [separate PR?] clean up `SingleTPUPlugin.save()` which is not used/needed anymore as its functionality moved to `XLACheckpointIO`.

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/plugins/training_type/single_tpu.py [hunk 1]
```diff
@@ -75,9 +75,6 @@ def pre_dispatch(self, trainer: "pl.Trainer") -> None:
         self.tpu_local_core_rank = xm.get_local_ordinal()
         self.tpu_global_core_rank = xm.get_ordinal()
 
-    def save(self, state_dict: Dict, path: _PATH) -> None:
-        xm.save(state_dict, path)
-
```
Comment (ananthsub) [inline pytorch_lightning/plugins/training_type/single_tpu.py:80]: nit: this could be a separate PR

File: pytorch_lightning/trainer/trainer.py [hunk 2]
```diff
@@ -1700,10 +1700,6 @@ def world_size(self) -> int:
         # some training types define a world size
         return getattr(self.training_type_plugin, "world_size", 1)
 
-    @property
-    def should_rank_save_checkpoint(self) -> bool:
-        return self.training_type_plugin.should_rank_save_checkpoint
-
```
Comment (ananthsub) [inline pytorch_lightning/trainer/trainer.py:1706]: since this is publicly exposed, this needs to go through a deprecation process

File: pytorch_lightning/callbacks/model_checkpoint.py [hunk 3]
```diff
@@ -600,9 +600,6 @@ def __resolve_ckpt_dir(self, trainer: "pl.Trainer") -> None:
 
         self.dirpath = ckpt_path
 
-        if not trainer.fast_dev_run and trainer.training_type_plugin.should_rank_save_checkpoint:
```
Comment (ananthsub) [inline pytorch_lightning/callbacks/model_checkpoint.py:603]: to confirm, if the trainer is in fast_dev_run, there should be no checkpoint directory created?

File: pytorch_lightning/plugins/training_type/single_tpu.py (hunk 1 above)
Comment (jjenniferdai) [reply pytorch_lightning/plugins/training_type/single_tpu.py:80]: noted in summary. it felt minor and relevant but happy to pull out to a separate PR

File: pytorch_lightning/trainer/trainer.py (hunk 2 above)
Comment (jjenniferdai) [reply pytorch_lightning/trainer/trainer.py:1706]: noted in summary with context

File: pytorch_lightning/callbacks/model_checkpoint.py (hunk 3 above)
Comment (jjenniferdai) [reply pytorch_lightning/callbacks/model_checkpoint.py:603]: yep it'll just skip checkpointing (so IOPlugin makedirs) too https://github.com/PyTorchLightning/pytorch-lightning/blob/ed84cef3afa8db1381162cf86bf2992bce71f9fb/pytorch_lightning/callbacks/model_checkpoint.py#L383-L391

File: pytorch_lightning/trainer/trainer.py (hunk 2 above)
Comment (ananthsub) [reply pytorch_lightning/trainer/trainer.py:1706]: thanks for adding the context! i believe the initial removal was incorrect, and we should honor the deprecation process

**Review by ananthsub — APPROVED** [review]

**Review by carmocca — APPROVED** [review]

File: pytorch_lightning/trainer/trainer.py (hunk 2 above)
Comment (carmocca) [reply pytorch_lightning/trainer/trainer.py:1706]: Resolving to merge as this PR is just addressing the merge mistake from #9901. Open an issue or comment on #9433 for this.
